### ISSUE TITLE
Inconsistent commit hash for git source

### ISSUE THREAD

**Issue by tpgillam** [astral-sh/uv#8050]
uv 0.4.20 -- we've run into the following scenario with some internal packages:

- `pyproject.toml` has a `[tool.uv.sources]` section that defines a git source for a package. Let's call it `moo`. It's pinned to commit hash `A`
- `uv.lock` is committed, and specifies commit hash `B` (not equal to `A`) for `moo`
- in a repository without a `.venv`, running `uv sync --locked` will install commit hash `B`

This inconsistency between `pyproject.toml` and `uv.lock` arose due to a bit of a git merge accident we think. But it would be nice if `uv` errored if it spots this issue, as then our CI would have caught it!

(I can construct a reproducing example if useful, but maybe it's obvious enough given description above - let me know if not!)

**Comment by charliermarsh** [comment]
Yeah that seems wrong... So the `pyproject.toml` said `A`, the lock said `B`, but `uv sync --locked` succeeded?

Just to confirm: you're using `moo = { git = "...", rev = "<full commit hash>" }`?

**Comment by charliermarsh** [comment]
As in: I locked, then changed the hash in `pyproject.toml` and see:

```
❯ uv sync --locked
Resolved 4 packages in 57ms
error: The lockfile at `uv.lock` needs to be updated, but `--locked` was provided. To update the lockfile, run `uv lock`.
```

**Comment by tpgillam** [comment]
OK... so it seems this `uv.lock` is also internally inconsistent!

Condensing, we have this situation:

`pyproject.toml`
```toml
[project]
name = "puffin"
dependencies = ["moo"]

[tool.uv.sources]
moo = { git = "https://github.com/MonumoLtd/moo.git", rev="A" }
```

`uv.toml`
```toml
[[package]]
name = "puffin"
[package.metadata]
requires-dist = [{ name = "moo", git = "https://github.com/MonumoLtd/moo.git?rev=A" }]

...

[[package]]
name = "moo"
source = { git = "https://github.com/MonumoLtd/moo#B" }
```

If I do `uv sync --locked` in a directory called `puffin` with these two files (admittedly populated with actual package names and commit hashes), then it creates a new `.venv` successfully, and in the output from uv I see

```
 + moo==0.1.0 (from git+https://github.com/MonumoLtd/moo.git@B)

```

**Comment by charliermarsh** [comment]
Any idea _how_ the `uv.lock` ended up in that state?

**Comment by tpgillam** [comment]
It looks to me like the original state was `moo` unpinned in pyproject. 

Then a PR (1) was merged that added a pin to A and correctly used A everywhere in the lock file.

Then a second PR (2) introduced the error - this PR has been in flight for a while, so it had merged back in from main after (1) had been merged. I can't tell if someone had accidentally mis-resolved a conflict in the lock file. But this is my best guess.

Anyway, certainly I agree this is a bit of a weird one and not of `uv`'s causing. We have lock files committed for certain projects to help aid reproducibility, which I suspect opens us up to this kind of fragility..

**Comment by charliermarsh** [comment]
Reproduced, thank you! Agree there's some odd edit happening outside of uv's control here, but I think we should be erroring and not silently accepting it.
